### PR TITLE
Add TLS communication between Galadriel Server and Harvester

### PR DESCRIPTION
**Pull request by maxlambrecht**
**Pull request check list**

- [X] Proper tests/regressions included?
- [ ] Documentation updated?

**Affected functionality**
N/A

**Description of change**

Galadriel Server exposes a TCP endpoint using TLS. The TLS certificate is generated by the `X509CA` disk implementation using a ROOT CA read from files in disk. The TLS certificate is automatically rotated every hour and only kept in memory. 

The Harvester connects to the Server using TLS. The TLS server certificate is validated using a configured trust bundle. The TLS client validates that the cert presented by the server have "galadriel-server" as a name.

A few improvements and refactors were done.

**Which issue this pull requests fixes**

Fixes [GLCP-39466](https://hpe.atlassian.net/browse/GLCP-39466).

### CODE REVIEW

File: cmd/harvester/cli/config.go [hunk 1]
```diff
@@ -28,6 +29,7 @@ type Config struct {
 type harvesterConfig struct {
 	SpireSocketPath       string `hcl:"spire_socket_path"`
 	ServerAddress         string `hcl:"server_address"`
+	TrustBundlePath       string `hcl:"trust_bundle_path"`
```
Comment (mchurichi) [inline cmd/harvester/cli/config.go:32]: Isn't the term `trust_bundle` a bit misleading in the context of the Harvester, given that it sits next to the spire server which has a trust bundle too? How about something like `server_ca_bundle` or similar? also, prefixing all spire related properties with `spire` and galadriel server properties with `server` might help

File: pkg/harvester/client/server.go [hunk 2]
```diff
@@ -147,6 +160,28 @@ func (c *client) PostBundle(ctx context.Context, req *common.PostBundleRequest)
...
 }
 
+func createTLSClient(trustBundlePath string) (*http.Client, error) {
+	caCert, err := os.ReadFile(trustBundlePath)
+	if err != nil {
+		return nil, err
+	}
+
+	caCertPool := x509.NewCertPool()
+	caCertPool.AppendCertsFromPEM(caCert)
```
Comment (mchurichi) [inline pkg/harvester/client/server.go:170]: Shouldn't we fail (or at least log a warning) if for some reason no certs were appended?

File: pkg/server/endpoints/run.go [hunk 3]
```diff
@@ -134,3 +183,73 @@ func (e *Endpoints) addTCPHandlers(server *echo.Echo) {
...
 }
+
+func (t *certificateSource) setTLSCertificate(cert *tls.Certificate) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	t.cert = cert
+}
+
+func (t *certificateSource) getTLSCertificate() *tls.Certificate {
+	t.mu.Lock()
```
Comment (mchurichi) [inline pkg/server/endpoints/run.go:194]: RLock?

File: pkg/server/endpoints/run.go [hunk 4]
```diff
@@ -134,3 +183,73 @@ func (e *Endpoints) addTCPHandlers(server *echo.Echo) {
...
+	e.Logger.Info("Starting TLS certificate rotator")
+
+	// Start a ticker that rotates the certificate every default interval
+	certRotationInterval := defaultTTL / 2
+	ticker := time.NewTicker(certRotationInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			e.Logger.Info("Rotating GCA TLS certificate")
```
Comment (mchurichi) [inline pkg/server/endpoints/run.go:209]:
```suggestion
			e.Logger.Info("Rotating Server TLS certificate")
```

File: pkg/server/endpoints/run.go [hunk 5]
```diff
@@ -134,3 +183,73 @@ func (e *Endpoints) addTCPHandlers(server *echo.Echo) {
...
+			cert, err := e.getTLSCertificate(ctx)
+			if err != nil {
+				errChan <- fmt.Errorf("failed to rotate GCA TLS certificate: %w", err)
```
Comment (mchurichi) [inline pkg/server/endpoints/run.go:212]:
```suggestion
				errChan <- fmt.Errorf("failed to rotate Server TLS certificate: %w", err)
```

File: pkg/server/endpoints/run.go [hunk 6]
```diff
@@ -134,3 +183,73 @@ func (e *Endpoints) addTCPHandlers(server *echo.Echo) {
...
+			}
+			e.certsStore.setTLSCertificate(cert)
+		case <-ctx.Done():
+			e.Logger.Info("Stopped GCA TLS certificate rotator")
```
Comment (mchurichi) [inline pkg/server/endpoints/run.go:216]:
```suggestion
			e.Logger.Info("Stopped Server TLS certificate rotator")
```

File: cmd/harvester/cli/config.go (hunk 1 above)
Comment (jufantozzi) [reply cmd/harvester/cli/config.go:32]: +1

File: pkg/harvester/client/server.go (hunk 2 above)
Comment (maxlambrecht) [reply pkg/harvester/client/server.go]: Absolutely. Good catch.

File: pkg/harvester/client/server.go [hunk 7]
```diff
@@ -147,6 +160,32 @@ func (c *client) PostBundle(ctx context.Context, req *common.PostBundleRequest)
...
+			RootCAs:    caCertPool,
+			ServerName: constants.GaladrielServerName,
+		},
+	}
+
+	client := &http.Client{
+		Transport: tr,
+	}
+
+	return client, nil
```
Comment (Victorblsilveira) [inline pkg/harvester/client/server.go:186]:
```suggestion
	return &http.Client{
		Transport: &http.Transport{
		 TLSClientConfig: &tls.Config{
			RootCAs:    caCertPool,
			ServerName: constants.GaladrielServerName,
		},
	}, nil
```

Comment (maxlambrecht) [reply pkg/harvester/client/server.go]: I agree refactored it to:

```
tr := &http.Transport{
	TLSClientConfig: &tls.Config{
		RootCAs:    caCertPool,
		ServerName: constants.GaladrielServerName,
	},
}

return &http.Client{
	Transport: tr,
}, nil
```
I don't like to nest more than two struct literals, it becomes hard to read.

**Review by jufantozzi — APPROVED** [review]
🏅

**Pull request merged**